src/main/java/com/microsoft/aad/msal4j/MsalServiceExceptionFactory.java

```diff
@@ -30,25 +30,26 @@ static MsalServiceException fromHttpResponse(HTTPResponse httpResponse){
         errorResponse.statusCode(httpResponse.getStatusCode());
         errorResponse.statusMessage(httpResponse.getStatusMessage());
 
-        if(errorResponse.error() != null &&
+        if (isRefreshTokenGrantRequest &&
```

Comment:
I don't know if it makes sense to just limit throwing InteractionRequiredError for just RefreshTokenGrant. What about in the case of username password and integrated windows authentication? interaction required is just as relevant in those cases as it is for refresh flow.

Reply:
both cases are "interactive", if it fails It means that provided credentials are wrong, there is no  plan B - plan B exists only for silent flow

Reply:
Not sure I would consider IWA and U/P interactive. To me it seems that an interactive request is one where the IDP renders UI, which is not the case for either of those. 

Also IWA and U/P can fail for more than just providing wrong credentials - one example is MFA. In this case, someone doing U/P could theoretically fall back to doing an interactive flow to meet the MFA requirement. I agree that it probably does not make sense to do this in an application, but anyways the InteractionRequired error is still 100% valid and the most transparent error to expose to an user. In the case that we made the change you are recommending, and someone is using one of this flows with MFA enabled on the tenant, what kind of error would they see?

Reply:
interactive vs non-interactive/silent distinguishing does make sense only for flows where user perform authorization "interactively" - like auth code flow - user redirected to identity provider authorization endpoint and he interact with it - so identity provider can give him additional challenges and he can satisfy it. In case of U/P or WIA - if credentials are wrong or missing there is not much we can do, we do not have plan B (different api) to handle it, like we do with auth code flow

Reply:
This is a great point @sangonzal,
I have seen many client applications implementing flow as:

try
  - AcquireTokenSilent
catch
   - IWA
catch 
    - Interactive

In fact, the spec states

`... for silent flow some errors of WS-Trust flow should go in UI required cache as well as errors from token endpoint. Applications have tendency to loop on WS-Trust flow, which would result in DDoS of UserRealm endpoint. There is no way such cases for the application to proceed without showing UI. Here is the list of these errors:

account is managed/unknown,
integration is not WS-Trust (SAML2.0)
missing MEX endpoint,
invalid MEX document,
MEX doesn’t have Windows Transport Endpoint,
failed to get SAML token from ADFS,
invalid_grant,
interaction_required.`

Reply:
So, you are suggesting to cache all invalid_grant responses?
What about next IWA  scenario: 
1. user get invalid_grant  because kerboros ticket expired.
2. Interaction required exception is thrown.
3. user execute kinit command to get new kerboros ticket.
4. user run IWA again and get cached exception 
5. cached exception will be returned for 2 mins, even though you have 
valid just generated kerberos ticket.
6. worst part about it  - That user is not even informed that he got cached response.

This is just one example of consequences of having dishonest/non_transparent API.

That is why i think caching of responses should be avoided / limited to some well considered specific cases.

Abuse of endpoints for WS-Trust flow is handled by Throttling instructions like Retry_Header, 429, 5xx

Reply:
Regarding business scope of MsalInteractionRequiredException:
- I believe that idea behind this exception is to inform that refresh token can not be used, 
so you need to provide credentials.
- if you already sending credentials but they are wrong - What does this message deliver to user ? From my perspective  MSALServiceException with "invalid_grant" error code should be used in this case

Reply:
I believe MSALs throw this exception for most (but not all) InvalidGrant and InteractionRequired errors. See this spec for details https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1148 (I think @sangonzal implemented this in Java lib). 

So this exception is not exclusive to `AcquireTokenSilent`, but to all silent flows: 

- AquireTokenSilent
- AcquireTokenInteractive when using a silent web ui
- IWA
- ROPC

As for the IWA scenario you described, it looks more like a corner case. Are you sure AAD throws UiRequiredException there?

Reply:
I sent an email to Sasha for clarifications

Reply:
"One of common status codes returned from authentication libraries in silent mode today is InvalidGrant. This status code means that the application should call the authentication library again, but in interactive mode. " - sounds pretty straight forward -> it is about silent mode, only one type of requests can be sent in silent mode - refresh token grant

Reply:
throttling design also explicitly highlight that scope of responses caching is silent requests.

Reply:
"it is about silent mode, only one type of requests can be sent in silent mode - refresh token grant" I don't agree with you here. The most meaningful difference IMO between "silent mode" and "non-silent mode" is showing UI. From that perspective the other flows are silent as well. 

InteractionRequired is a valid error to expose for the other flows. What would we expose to users in the case of MFA for example (which is relevant in all flows)? InteractionRequired (pulling up UI and having someone click on it) is the way to resolve that issue, and the most transparent exception to expose to users, whether it's actionable or not for them. 

From the perspective of throttling, wouldn't we also want to throttle applications stuck in a loop in ROPC or IWA flows if they are getting interaction required returned from the server?

Reply:
One more thing to consider is that changing this behavior (only throwing InteractionRequired for silent requests) is a breaking change, and would require bumping up the major version.

Reply:
"InteractionRequired is a valid error to expose for the other flows." - even for non interactive  flows ?

Reply:
removing redefinition of InteractionRequired Exception from the scope of this PR